File: solutions/zig/pangram/1/pangram.zig
// Your task is to figure out if a sentence is a pangram.
//
// A pangram is a sentence using every letter of the alphabet at least once.
// It is case insensitive, so it doesn't matter if a letter is lower-case (e.g. `k`) or upper-case (e.g. `K`).
//
// For this exercise, a sentence is a pangram if it contains each of the 26 letters in the English alphabet.

const std = @import("std");

pub fn isPangram(str: []const u8) bool {
    var letters: [26]bool = .{false} ** 26;
    for (str) |char| {
        if (std.ascii.isAlphabetic(char)) {
            const index = std.ascii.toLower(char) - 'a';
            letters[index] = true;
        }
    }

    for (letters) |state| {
        if (!state) return false;
    }

    return true;
}

File: solutions/c/pangram/1/pangram.c
#include "pangram.h"
#include <stdlib.h>

static int get_char_code(char c) {
  if (c >= 'a' && c <= 'z') return c - 'a';
  else return c - 'A';
}

static const int ALPHABET_SIZE = 26;

bool is_pangram(const char *sentence) {
  if (sentence == NULL) return false;

  int mask = 0;

  for (char *c = (char *)sentence; *c; c++) {
    int code = get_char_code(*c);
    if (code >= 0 && code < ALPHABET_SIZE) mask |= 1 << code;
  }

  return (mask + 1) == (1 << ALPHABET_SIZE);
}
